### PR TITLE
chore(core): implement concurrent symbol index reader

### PR DESCRIPTION
**Pull request by puzpuzpuz**
New symbol index reader supports thread-safe cursors that don't remap key and value memory on access. The assumption
is that the owner is responsible for calling of() method before it starts accessing the cursors.

Symbol map reader class was migrated to the new index reader as a step of the async offload symbol filters thread safety fix (see #2107).

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 98 / 106 (92.45%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ConcurrentBitmapIndexFwdReader.java|77|85|90.59%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexUtils.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|2|2|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 97 / 105 (92.38%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ConcurrentBitmapIndexFwdReader.java|77|85|90.59%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexUtils.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 97 / 105 (92.38%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ConcurrentBitmapIndexFwdReader.java|77|85|90.59%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexUtils.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|1|1|100.00%|

**Review by bluestreak01 — APPROVED** [review]

File: core/src/main/java/io/questdb/cairo/ConcurrentBitmapIndexFwdReader.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    public RowCursor initCursor(RowCursor rowCursor, int key, long minValue, long maxValue) {
+        Cursor cursor = null;
+        if (rowCursor != null && rowCursor != EmptyRowCursor.INSTANCE) {
+            cursor = (Cursor) rowCursor;
+            assert cursor.parentReader() == this;
+        }
+
+        if (key == 0 && unIndexedNullCount > 0 && minValue < unIndexedNullCount) {
+            // we need to return some nulls and the whole set of actual index values
+            if (cursor == null) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/ConcurrentBitmapIndexFwdReader.java:89]: this is untested and is quite critical

File: core/src/main/java/io/questdb/cairo/ConcurrentBitmapIndexFwdReader.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+
+        if (key < keyCount) {
+            if (cursor == null) {
+                cursor = getCursor(false);
+            }
+            cursor.of(key, minValue, maxValue, keyCount, 0, 0);
+            return cursor;
+        }
+
+        return EmptyRowCursor.INSTANCE;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/cairo/ConcurrentBitmapIndexFwdReader.java:104]: so is this

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Review by ideoma — DISMISSED** [review]

File: core/src/main/java/io/questdb/cairo/ConcurrentBitmapIndexFwdReader.java (hunk 1 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/cairo/ConcurrentBitmapIndexFwdReader.java:76]: Done in 9cc86ee6e93baaf2ac4b7207b45d486d66dccf07

File: core/src/main/java/io/questdb/cairo/ConcurrentBitmapIndexFwdReader.java (hunk 2 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/cairo/ConcurrentBitmapIndexFwdReader.java:91]: Done in 9cc86ee6e93baaf2ac4b7207b45d486d66dccf07

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 93 / 96 (96.88%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/ConcurrentBitmapIndexFwdReader.java|73|76|96.05%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexUtils.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|1|1|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Comment by puzpuzpuz** [comment]
@ideoma @bluestreak01 thanks for your reviews!

**Pull request merged**